File: backend/internal/domain/group.go
package domain

import "time"

type Group struct {
	Id           int
	Name         string
	Description  *string
	Enabled      bool
	DateAdded    time.Time
	DateModified time.Time
}

type GroupSet struct {
	Groups []Group
}

type AddGroupCommand struct {
	Name        string
	Description *string
	Enabled     bool
}

type UpdateGroupCommand struct {
	Name        string
	Description *string
	Enabled     *bool
}

type RemoveGroupCommand struct {
	Name string
}

File: frontend/src/types/group.ts
import type { PiholeNodeRef } from './pihole';

export type Group = {
	id: number;
	name: string;
	description: string | null;
	enabled: boolean;
	dateAdded: string;
	dateModified: string;
};

// Per-node result

type ListGroupsNode = {
	node: PiholeNodeRef;
	groups: Group[];
	error?: string;
};

type ListGroupsSummary = {
	totalNodes: number;
	okNodes: number;
	errorNodes: number;
};

export type ListGroupsResponse = {
	summary: ListGroupsSummary;
	nodes: Record<string, ListGroupsNode>;
};

type GroupMutateNode = {
	node: PiholeNodeRef;
	groups: Group[];
	error?: string;
};

export type GroupMutateResponse = {
	nodes: Record<string, GroupMutateNode>;
};

type RemoveGroupNode = {
	node: PiholeNodeRef;
	removed: boolean;
	error?: string;
};

type RemoveGroupSummary = {
	total: number;
	removed: number;
	failed: number;
};

export type RemoveGroupResponse = {
	summary: RemoveGroupSummary;
	nodes: Record<string, RemoveGroupNode>;
};
